groups/memory_group_service.go

```diff
@@ -0,0 +1,186 @@
+package groups
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/topfreegames/pitaya/config"
+	"github.com/topfreegames/pitaya/constants"
+)
+
+var (
+	memoryGroupsMu sync.RWMutex
+	memoryGroups   map[string]*MemoryGroup
+	memoryOnce     sync.Once
+)
+
+// MemoryGroupService base in server memory solution
+type MemoryGroupService struct {
+}
+
+// MemoryGroup is the struct stored in each group key(which is the name of the group)
+type MemoryGroup struct {
+	Uids        []string
+	LastRefresh int64
+	TTL         int64
+}
+
+// NewMemoryGroupService returns a new group instance
+func NewMemoryGroupService(conf *config.Config) *MemoryGroupService {
+	memoryOnce.Do(func() {
+		memoryGroups = make(map[string]*MemoryGroup)
+		go func() {
+			for now := range time.Tick(conf.GetDuration("pitaya.groups.memory.tickduration")) {
+				memoryGroupsMu.Lock()
+				for groupName, mg := range memoryGroups {
+					if mg.TTL != 0 && now.UnixNano()-mg.LastRefresh > mg.TTL {
+						delete(memoryGroups, groupName)
+					}
+				}
+				memoryGroupsMu.Unlock()
+			}
+		}()
+	})
+	return &MemoryGroupService{}
+}
+
+func getMemoryGroup(groupName string) (*MemoryGroup, error) {
+	memoryGroupsMu.Lock()
+	defer memoryGroupsMu.Unlock()
+
+	mg, ok := memoryGroups[groupName]
+	if !ok {
+		return nil, constants.ErrGroupNotFound
+	}
+
+	return mg, nil
+}
+
+// GroupCreate creates a group without TTL
+func (c *MemoryGroupService) GroupCreate(ctx context.Context, groupName string) error {
+	memoryGroupsMu.Lock()
+	defer memoryGroupsMu.Unlock()
+
+	if _, ok := memoryGroups[groupName]; ok {
+		return constants.ErrGroupAlreadyExists
+	}
+
+	memoryGroups[groupName] = &MemoryGroup{}
+	return nil
+}
+
+// GroupCreateWithTTL creates a group with TTL, which the go routine will clean later
+func (c *MemoryGroupService) GroupCreateWithTTL(ctx context.Context, groupName string, ttlTime time.Duration) error {
+	memoryGroupsMu.Lock()
+	defer memoryGroupsMu.Unlock()
+
+	if _, ok := memoryGroups[groupName]; ok {
+		return constants.ErrGroupAlreadyExists
+	}
+
+	memoryGroups[groupName] = &MemoryGroup{LastRefresh: time.Now().UnixNano(), TTL: ttlTime.Nanoseconds()}
+	return nil
+}
+
+// GroupMembers returns all member's UID in given group
+func (c *MemoryGroupService) GroupMembers(ctx context.Context, groupName string) ([]string, error) {
+	mg, err := getMemoryGroup(groupName)
+	if err != nil {
+		return nil, err
+	}
+
+	return mg.Uids, nil
```

Comment:
Here you return the array of ids, which is an array, any function that uses the return value of this function is able to modify the group elements, you should ideally return a copy of the array

Reply:
Thanks for the advice! Changed the implementation and copied the slice values to another one!